File: dsa_algo/leetcode/2_pointers/assorted/c++/15-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
       std::vector<std::vector<int>> result;
       sort(nums.begin(), nums.end());
       for(int i = 0; i < nums.size(); i ++)
       {
           if(i > 0 && nums[i] == nums[i - 1]) //check if value already used 
                continue;    //if yes, skip to next iteration
            int left = i + 1;
            int right = nums.size() - 1;
            while(left < right)
            {
                int threesome = nums[i] + nums[left] + nums[right];
                if(threesome > 0)
                    right--;
                else if(threesome < 0)
                    left++;
                else
                {
                    std::vector<int> sub_result;
                    sub_result.push_back(nums[i]);
                    sub_result.push_back(nums[left]);
                    sub_result.push_back(nums[right]);
                    result.push_back(sub_result);
                    left++;
                    while(nums[left] == nums[left - 1] && left < right )
                        left++;
                }
            }
       }
       return result;
    }
};

File: dsa_algo/leetcode/ts/src/2_pointers/15-3sum.ts
const threeSum = (nums: number[]): number[][] =>{
    const result: number [][] = []; 
    nums.sort((a, b) => a - b);
    for(let i = 0; i < nums.length; i++)
    {
       if(i > 0 && nums[i] == nums[i - 1])
          continue;
       let left = i + 1;
       let right = nums.length - 1;
       while( left < right)
       {
          let threesum = nums[i] + nums[left] + nums[right];
          if (threesum > 0)
             right --;
          else if(threesum < 0)
             left++;
          else
          {
             let sub_result: number [] = [];
             sub_result.push(nums[i]);
             sub_result.push(nums[left]);
             sub_result.push(nums[right]);
             result.push(sub_result);
             left ++;
             while(nums[left] == nums[left - 1] && left < right)
                left++;
          }
 
       }
 
    }
    return result;
 };